src/metabase/api/collection.clj

```diff
@@ -187,18 +207,40 @@
   [id]
   (collection-detail (api/read-check Collection id)))
 
+(defn- coerce-vec [to-coerce]
+  (if (and (some? to-coerce) (not (vector? to-coerce))) [to-coerce] to-coerce))
+
 (api/defendpoint GET "/:id/items"
   "Fetch a specific Collection's items with the following options:
 
-  *  `model` - only include objects of a specific `model`. If unspecified, returns objects of all models
-  *  `archived` - when `true`, return archived objects *instead* of unarchived ones. Defaults to `false`."
-  [id model archived]
-  {model    (s/maybe (apply s/enum valid-model-param-values))
-   archived (s/maybe su/BooleanString)}
-  (collection-children
-    (api/read-check Collection id)
-    {:model     (keyword model)
-     :archived? (Boolean/parseBoolean archived)}))
+  *  `models` - only include objects of a specific set of `models`. If unspecified, returns objects of all models
+  *  `archived` - when `true`, return archived objects *instead* of unarchived ones. Defaults to `false`.
+  *  `pinned_state` - when `is_pinned`, return pinned objects only.
+                   when `is_not_pinned`, return non pinned objects only.
+                   when `all`, return everything.
+  *  `limit` - limit for pagination.
+  *  `offset` - offset for pagination."
+  [id models archived pinned_state limit offset]
+  {models       (s/maybe models-schema)
+   archived     (s/maybe su/BooleanString)
+   pinned_state (s/maybe (apply s/enum valid-pinned-state-values))
+   limit        (s/maybe su/IntStringGreaterThanZero)
+   offset       (s/maybe su/IntStringGreaterThanOrEqualToZero)}
+  (api/check-valid-page-params limit offset)
+  (let [model-kwds    (apply hash-set (map keyword (coerce-vec models)))
+        children-res  (collection-children (api/read-check Collection id)
+                                           {:models       model-kwds
+                                            :archived?    (Boolean/parseBoolean archived)
+                                            :pinned-state (keyword pinned_state)})
+        limit-int     (some-> limit Integer/parseInt)
+        offset-int    (some-> offset Integer/parseInt) ]
+    {:data   (cond->> (vec children-res)
+               (some? offset-int) (drop offset-int)
```

Comment:
Does `offset` make more sense than `page`? I think `page` is a much more common way of doing paging in REST APIs. Doesn't make a big difference to me, just curious

Reply:
`offset limit` has different tradeoffs versus using the like 4 other ways of paging in an endpoint. so, like, nearly all usage in b2c land is slurping from the infinite bowl, in which case you want to be using another method of pagination. my guess is that we are pretty much just not in b2c land, so I chose limit offset which has big disadvantages but big advantages (eg, you can go to an arbitrary page, which you just can't with nearly all other pagination methods)

Reply:
also, everyone and their mother supports offset limit